p2p/src/lib.rs

```diff
@@ -101,3 +104,60 @@ where
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::net::libp2p::Libp2pService;
+    use crate::pubsub::PubSubManager;
+    use crate::swarm::SwarmManager;
+    use common::chain::config;
+    use libp2p::Multiaddr;
+    use tokio::sync::mpsc;
+    use crate::sync::SyncManager;
+
+    #[tokio::test]
+    async fn test_subsys_init() {
+        let addr: Multiaddr = test_utils::make_address("/ip6/::1/tcp/");
+        let config = Arc::new(config::create_mainnet());
+        let (conn, pubsub, sync) = Libp2pService::start(
+            addr,
+            &[],
+            &[],
+            Arc::clone(&config),
+            std::time::Duration::from_secs(10),
+        )
+        .await
+        .unwrap();
+
+        let storage = blockchain_storage::Store::new_empty().unwrap();
+        let mut manager = subsystem::Manager::new("mintlayer");
+        let consensus = manager.add_subsystem(
+            "consensus",
+            consensus::make_consensus(config::create_mainnet(), storage.clone()).unwrap(),
+        );
+
+		// NOTE: these channels are needed because `SwarmManager` must be able to call
+		// `SyncManager` and `PubSubManager` and the managers cannot be initialized if
+		// handles to them are acquired only after they are created.
+        let (tx1, rx1) = mpsc::channel(16);
+        let (tx2, rx2) = mpsc::channel(16);
+
+        let mut swarm_mgr = SwarmManager::<Libp2pService>::new(Arc::clone(&config), conn, rx1, tx2);
+        let swarm_handle = manager.add_raw_subsystem("swarm", |call_rq, shut_rq| async move {
+            swarm_mgr.run(call_rq, shut_rq).await
+        });
+
+        let mut pubsub_mgr = PubSubManager::<Libp2pService>::new(pubsub, consensus, swarm_handle.clone());
+        let pubsub = manager.add_raw_subsystem("pubsub", |call_rq, shut_rq| async move {
+            pubsub_mgr.run(call_rq, shut_rq).await
+        });
+
+        let mut sync_mgr = SyncManager::<Libp2pService>::new(sync, rx2, swarm_handle);
+        let sync = manager.add_raw_subsystem("sync", |call_rq, shut_rq| async move {
+            sync_mgr.run(call_rq, shut_rq).await
+        });
+
+        manager.main().await;
+    }
```

Comment:
@iljakuklic 

This is where I hit a bit of roadblock. The managers would need handles passed to each other but I think it's still not possible and I'm not sure if it will ever be because maybe it's too hard to/not worth the effort to decouple the handle creation from spawning the subsystem function.

Channels work since I can create them before I create the managers and I can implement some nice `Handle<T>`-style interface for them with macros.

I'll think about the relationship between them a little bit and maybe some pattern will emerge that doesn't require passing the handle to all of them. Right now the "worst case" (that handles can't be used between P2P objects) is not the end of the world because channels work just fine.

Reply:
I'd like to extend the subsystem framework with an event publish-subscribe mechanism at some point. Maybe that could be used? Depends what the interaction between subsystems looks like.

Reply:
I had a chat with Sam and it seems that using subsystems wrapper for individual manager objects inside P2P is not correct behavior so you can ignore this PR.